setup/main.go

```diff
@@ -0,0 +1,1653 @@
+		SELECT string_agg(quote_literal(param) || '::unknown', ',') FROM unnest(params) p(param) INTO prepared_params;
+
+		EXECUTE 'PREPARE pganalyze_explain AS ' || prepared_query;
+		BEGIN
+			EXECUTE 'EXPLAIN (VERBOSE, FORMAT JSON) EXECUTE pganalyze_explain(' || prepared_params || ')' INTO STRICT result;
+		EXCEPTION WHEN OTHERS THEN
+			DEALLOCATE pganalyze_explain;
+			RAISE;
+		END;
+		DEALLOCATE pganalyze_explain;
+	ELSE
+		EXECUTE 'EXPLAIN (VERBOSE, FORMAT JSON) ' || prepared_query INTO STRICT result;
+	END IF;
+
+	RETURN result;
+END
+$$ LANGUAGE plpgsql VOLATILE SECURITY DEFINER;`)
+	},
+}
+
+var checkAutoExplainAvailable = &Step{
+	Description: "Prepare for auto_explain install",
+	Check: func(state *SetupState) (bool, error) {
+		if state.SkipAutomatedExplain {
+			return true, nil
+		}
+		logExplain, err := usingLogExplain(state.CurrentSection)
+		if err != nil || logExplain {
+			return logExplain, err
+		}
+		err = state.QueryRunner.Exec("LOAD 'auto_explain'")
+		if err != nil {
+			if strings.Contains(err.Error(), "No such file or directory") {
+				return false, nil
+			}
+
+			return false, err
+		}
+		return true, err
+	},
+	Run: func(state *SetupState) error {
+		// TODO: install contrib package?
+		return errors.New("module auto_explain is not available")
+	},
+}
+
+var enableAutoExplain = &Step{
+	Description: "Enable auto_explain",
+	Check: func(state *SetupState) (bool, error) {
+		if state.SkipAutomatedExplain {
+			return true, nil
+		}
+		logExplain, err := usingLogExplain(state.CurrentSection)
+		if err != nil || logExplain {
+			return logExplain, err
+		}
+		spl, err := getPendingSharedPreloadLibraries(state.QueryRunner)
+		if err != nil {
+			return false, err
+		}
+		return strings.Contains(spl, "auto_explain"), nil
+	},
+	Run: func(state *SetupState) error {
+		var doAdd bool
+		err := survey.AskOne(&survey.Confirm{
+			Message: "Add auto_explain to shared_preload_libraries (will be saved to Postgres)?",
+			Default: false,
+			Help:    "Postgres will have to be restarted in a later step to apply this configuration change; learn more about shared_preload_libraries here: https://www.postgresql.org/docs/current/runtime-config-client.html#GUC-SHARED-PRELOAD-LIBRARIES",
+		}, &doAdd)
+		if err != nil {
+			return err
+		}
+		if !doAdd {
+			return nil
+		}
+
+		existingSpl, err := getPendingSharedPreloadLibraries(state.QueryRunner)
+		if err != nil {
+			return err
+		}
+		var newSpl string
+		if existingSpl == "" {
+			newSpl = "auto_explain"
+		} else {
+			newSpl = existingSpl + ",auto_explain"
+		}
+		return applyConfigSetting("shared_preload_libraries", newSpl, state.QueryRunner)
+	},
+}
+
+var reloadCollector = &Step{
+	Description: "Reload collector configuration",
+	Check: func(state *SetupState) (bool, error) {
+		// N.B.: there's no way to tell whether the collector actually needs to reload; we
+		// force this once per setup helper invocation (this starts out as false and is set
+		// to true below)
```

Comment:
I was wondering why I was prompted again when cancelling the setup steps and re-running, but I think this works for now.

For reference, this is what I ran (I quit the setup whilst on the last question shown here):

```
? Continue with setup? Yes
? Select Postgres superuser to connect as for configuration purposes postgres
? Proceed to configuring optional Log Insights feature? Yes
? Proceed to configuring optional Automated EXPLAIN feature? Yes
? The collector configuration must be reloaded for changes to take effect; reload now? Yes
? Review auto_explain configuration settings? Yes
? Set auto_explain.log_analyze to (will be saved to Postgres):  [Use arrows to move, type to filter, ? for more help]
> on (recommended)
  off
```

Reply:
I suppose we could track whether we've made any changes to the collector config at all in the session--if not, there's no need to restart.